File: matrixMultiplication/MatrixMultiplicationTest.py
import time
import psutil  # For CPU usage tracking
from memory_profiler import memory_usage
from MatrixMultiplication import MatrixMultiplication

class MatrixMultiplicationTest:
    def __init__(self, matrix_sizes, runs_per_size=5):
        self.matrix_sizes = matrix_sizes
        self.runs_per_size = runs_per_size

    def get_used_memory(self):
        mem = memory_usage(-1, interval=0.1, timeout=1)
        return mem[0]

    def get_cpu_usage(self):
        return psutil.cpu_percent(interval=None)  # Returns CPU usage percentage

    def run_test(self):
        for size in self.matrix_sizes:
            print(f"=== Matrix Size: {size}x{size} ===")
            total_time = 0.0
            total_memory = 0.0
            total_cpu = 0.0  # To store cumulative CPU usage

            for run in range(1, self.runs_per_size + 1):
                mm = MatrixMultiplication(size)
                mm.initialize_matrices()
                memory_before = self.get_used_memory()
                cpu_before = self.get_cpu_usage()
                start_time = time.time()

                mm.multiply()

                end_time = time.time()
                memory_after = self.get_used_memory()
                cpu_after = self.get_cpu_usage()

                elapsed_time = end_time - start_time
                total_time += elapsed_time
                memory_used = memory_after - memory_before
                total_memory += memory_used
                cpu_used = cpu_after - cpu_before  # Difference in CPU usage
                total_cpu += cpu_used

                print(f"Run {run}: Time = {elapsed_time:.3f} seconds, Memory Used = {memory_used:.2f} MB, CPU Used = {cpu_used:.2f}%")

            average_time = total_time / self.runs_per_size
            average_memory = total_memory / self.runs_per_size
            average_cpu = total_cpu / self.runs_per_size

            print(f"Average: Time = {average_time:.3f} seconds, Average Memory Used = {average_memory:.2f} MB, Average CPU Used = {average_cpu:.2f}%\n")

if __name__ == "__main__":
    matrix_sizes = [256, 512, 1024, 2048]
    runs_per_size = 5

    tester = MatrixMultiplicationTest(matrix_sizes, runs_per_size)
    tester.run_test()

File: MatrixMultiplication/src/test/java/MatrixMultiplicationTest.java
import oshi.SystemInfo;
import oshi.hardware.CentralProcessor;
import java.io.FileWriter;
import java.io.IOException;

public class MatrixMultiplicationTest {

    public static void main(String[] args) {
        int[] matrixSizes = {256, 512, 1024, 2048};
        int runsPerSize = 5;

        // Initialize OSHI for CPU usage tracking
        SystemInfo si = new SystemInfo();
        CentralProcessor processor = si.getHardware().getProcessor();

        // Store the initial CPU ticks
        long[] previousTicks = processor.getSystemCpuLoadTicks();

        try (FileWriter writer = new FileWriter("benchmark/results.csv")) {
            writer.write("Matrix Size,Run,Time (seconds),Memory Used (MB),CPU Used (%)\n");

            for (int size : matrixSizes) {
                System.out.println("=== Matrix Size: " + size + "x" + size + " ===");
                double totalTime = 0.0;
                long totalMemory = 0;
                double totalCpu = 0.0;

                for (int run = 1; run <= runsPerSize; run++) {
                    MatrixMultiplication mm = new MatrixMultiplication(size);
                    mm.initializeMatrices();
                    long memoryBefore = getUsedMemory();

                    long startTime = System.nanoTime();
                    mm.multiply();
                    long endTime = System.nanoTime();

                    long memoryAfter = getUsedMemory();

                    // Get the CPU usage based on the previous ticks
                    double cpuUsed = processor.getSystemCpuLoadBetweenTicks(previousTicks) * 100;  // Returns CPU usage as a percentage
                    previousTicks = processor.getSystemCpuLoadTicks();  // Update previous ticks

                    // Calculate time and memory
                    double elapsedTime = (endTime - startTime) / 1e9;  // Convert to seconds
                    long memoryUsed = (memoryAfter - memoryBefore) / (1024 * 1024);  // Convert to MB

                    // Add to totals
                    totalTime += elapsedTime;
                    totalMemory += memoryUsed;
                    totalCpu += cpuUsed;

                    // Print and write results to file
                    System.out.printf("Run %d: Time = %.3f seconds, Memory Used = %d MB, CPU Used = %.2f%%%n", run, elapsedTime, memoryUsed, cpuUsed);
                    writer.write(String.format("%d,%d,%.3f,%d,%.2f\n", size, run, elapsedTime, memoryUsed, cpuUsed));
                }

                // Average results
                double averageTime = totalTime / runsPerSize;
                double averageMemory = (double) totalMemory / runsPerSize;
                double averageCpu = totalCpu / runsPerSize;
                System.out.printf("Average: Time = %.3f seconds, Average Memory Used = %.2f MB, Average CPU Used = %.2f%%%n%n", averageTime, averageMemory, averageCpu);
            }
        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }
    }

    // Method to get used memory in bytes
    private static long getUsedMemory() {
        Runtime runtime = Runtime.getRuntime();
        return runtime.totalMemory() - runtime.freeMemory();
    }
}
